planet/src/main/webapp/js/placesItinerary.js

```diff
@@ -13,56 +13,107 @@
+
+import TimeRange from './TimeRange.js';
+
+// Declare global functions.
 window.openAddPlaceForm = openAddPlaceForm;
 window.closeAddPlaceForm = closeAddPlaceForm;
 
+// Declare global variables/constants.
 const database = firebase.database();
+let map = new google.maps.Map(document.getElementById('empty-map'));
 
 export function renderPlaces() {
     const userId = firebase.auth().currentUser.uid;
     const placesRef = database.ref('users/' + userId + '/places');
     placesRef.once('value', (placesSnapshot) => {
         const placesContainer = document.getElementById('places');
         placesContainer.innerHTML = '';
-        placesSnapshot.forEach((childPlace) => {
-            let placeDetailsRef = database.ref('places/' + childPlace.val());
-            placeDetailsRef.once('value', (placeDetailsSnapshot) => {
-                let placeObject = placeDetailsSnapshot.val();
-                let placeElement = createPlaceElement (placeDetailsSnapshot.key,
-                                                placeObject.name, 
-                                                placeObject.address);
-                placesContainer.appendChild(placeElement);
-            });
-        });
+        const savedPlaces = placesSnapshot.val();
+        for (const placeId in savedPlaces) {
+            // Make request with fields
+            let placeRequest = {
+                placeId: placeId,
+                fields: ['place_id','name','rating','formatted_address',
+                    'opening_hours','photos','url']
+            };
+            let service = new google.maps.places.PlacesService(map);
+            service.getDetails(placeRequest, renderPlaceDetailsCallback); 
+        }
     });
 }
 
-function createPlaceElement(ref, name, address) {
+function renderPlaceDetailsCallback(place, status) {
+    if (status == google.maps.places.PlacesServiceStatus.OK) {
+        const placesContainer = document.getElementById('places');
+        const placeElement = createPlaceElement(place);
+        placesContainer.appendChild(placeElement);
+    }    
+}
+
+function createPlaceElement(place) {
     const placeElement = document.createElement('div');
     placeElement.setAttribute('class', 'card place yellow lighten-4');
-    placeElement.setAttribute('id', 'place-' + ref);
+    placeElement.setAttribute('id', 'place-' + place.place_id);
 
+    // Create the add button
     const addButton = document.createElement('div');
     addButton.setAttribute('class', 'place-button');
     
-    const eventFromThisPlace = document.getElementById(ref);
+    const eventFromThisPlace = document.getElementById(place.place_id);
     if (eventFromThisPlace) {
         addButton.innerHTML = `<i class="material-icons small">playlist_add_check</i>`;
     }else {
-        addButton.setAttribute('onclick', 'openAddPlaceForm(event, "' + ref + '")');
+        addButton.setAttribute('onclick', 'openAddPlaceForm(event, "' + place.place_id + '")');
```

Comment:
I'm trying to understand this logic. If this is a button on an existing place, you already have the place object. Could you store that object somewhere so that you don't need to load it from the places API when you click thebutton to add it to the itinerary?

One option would be to store it as a map of place id -> place object in the module's namespace
I'm sure there are other options as well.

Reply:
Do you mean when users are adding a place to the itinerary, we don't have to make another api call, instead we can just get the place information from the dom? (maybe assigning the elements ids like `place-name-placeid`, and `place-address-placeid`)?

Reply:
> Do you mean when users are adding a place to the itinerary, we don't have to make another api call,
Yes that's what I mean.

> maybe assigning the elements ids like place-name-placeid, and place-address-placeid
That sounds like a good option if you're planning to display all of that data. If not, storing them in a JS object is probably cleaner.

Reply:
Sounds good, I made the place objects :)